streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -519,107 +520,169 @@ private long computeLatency() {
+        if (records != null && !records.isEmpty()) {
+            int numAddedRecords = 0;
 
-                    for (TopicPartition partition : records.partitions()) {
-                        StreamTask task = activeTasksByPartition.get(partition);
-                        numAddedRecords += task.addRecords(partition, records.records(partition));
-                    }
-                    streamsMetrics.skippedRecordsSensor.record(records.count() - numAddedRecords, timerStartedMs);
-                    polledRecords = true;
-                } else {
-                    polledRecords = false;
-                }
+            for (TopicPartition partition : records.partitions()) {
+                StreamTask task = activeTasksByPartition.get(partition);
+                numAddedRecords += task.addRecords(partition, records.records(partition));
+            }
+            streamsMetrics.skippedRecordsSensor.record(records.count() - numAddedRecords, timerStartedMs);
+        }
+    }
 
-                // only record poll latency is long poll is required
-                if (longPoll) {
-                    streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
+    /**
+     * Schedule the records processing by selecting which record is processed next. Commits may
+     * happen as records are processed.
+     * @tasks The tasks that have records.
+     * @param recordsProcessedBeforeCommit number of records to be processed before commit is called.
+     *                                     if UNLIMITED_RECORDS, then commit is never called
+     * @return Number of records processed since last commit.
+     */
+    private long processAndPunctuate(final Map<TaskId, StreamTask> tasks,
+                                     final long recordsProcessedBeforeCommit) {
+
+        long totalProcessedEachRound;
+        long totalProcessedSinceLastCommit = 0;
+        // Round-robin scheduling by taking one record from each task repeatedly
+        // until no task has any records left
+        do {
+            totalProcessedEachRound = 0;
+            for (StreamTask task : tasks.values()) {
+                int numBuffered = task.process();
+                // we processed one record,
+                // and more are buffered waiting for the next round
+                if (numBuffered > 0) {
+                    totalProcessedEachRound++;
+                    totalProcessedSinceLastCommit++;
                 }
             }
+            if (recordsProcessedBeforeCommit != UNLIMITED_RECORDS &&
+                totalProcessedSinceLastCommit >= recordsProcessedBeforeCommit) {
+                totalProcessedSinceLastCommit = 0;
+                long processLatency = computeLatency();
+                streamsMetrics.processTimeSensor.record(processLatency / (double) totalProcessedSinceLastCommit,
+                    timerStartedMs);
+                maybeCommit(this.timerStartedMs);
```

Comment:
One meta question: since we are only checking for commit at the round boundary anyways (by `round boundary` I mean, suppose we have 5 tasks, we will not check committing when only processed task1,2,3.). And the checking condition of `maybeCommit` itself is minimal, could we just trigger `maybeCommit` every time we have done a round of one-record-per-task?

Also I'm wondering if we should update `this.timerStartedMs` once we have successfully executed the commit since otherwise we will always commit once every each round right?